api/envoy/config/listener/v3/listener.proto

```diff
@@ -103,8 +103,15 @@ message Listener {
 
   // The address that the listener should listen on. In general, the address must be unique, though
   // that is governed by the bind rules of the OS. E.g., multiple listeners can listen on port 0 on
+  // Linux as the actual port will be allocated by the OS. This is deprecated, using `addresses` instead.
+  core.v3.Address address = 2
+      [deprecated = true, (envoy.annotations.deprecated_at_minor_version) = "3.0"];
+
+  // A list of addresses that the listener should listen on. All the addresses in
+  // a listener should have same socket type (e.g TCP, UDP, internal). In general, the address must be unique, though
+  // that is governed by the bind rules of the OS. E.g., multiple listeners can listen on port 0 on
   // Linux as the actual port will be allocated by the OS.
-  core.v3.Address address = 2 [(validate.rules).message = {required: true}];
+  repeated core.v3.Address addresses = 32;
```

Comment:
I think it's valid to have the ports be different.

Would it make sense to allow both UDP and TCP to support TCP/QUIC on the same listener? cc @alyssawilk @danzh2010 ?

From a proto types point of view, I think it will be cleanest to continue using `core.v3.Address`, but add additional validation on the protocol matching, if we decide that is needed.

Reply:
I think we'd end up creating multiple listeners under the hood, but since most folks want HTTP/443 and UDP/443 to have roughly the same config, I think it'd be nice to not have to duplicate the full filter chain, HCM etc.
I think we'd have to do the same thing we do with upstream auto-config, and make sure that we configure a quic transport socket factory and TCP would use the "inner" TLS socket factory but there's a reference implementation and it shouldn't be tricky. I'm probably missing some other API gotchas though,  @danzh2010 thoughts?

Reply:
I add the check in the code https://github.com/envoyproxy/envoy/pull/19367/files#diff-b0d5e1a1317df8e22a556ee74d6f728379c5f85ce640a9404acc14bfe8192ad3R361, to ensure only one protocol type in single listener.

I thought the listener filter, like proxy protocol, which doesn't work with UDP? that is why I added the limit.

But yes, seems the HTTP layer should be same for quick and http

Reply:
QUIC only supports one L4 filter which is HCM. So IMO sharing L4 filter chain config might not be a good idea. And sharing HCM config is not easy as well. Let alone transport socket, HCM codec_type is also hard to share. Currently `auto` means checking the first few bytes of TCP payload to decide. 

I think config sharing between H2 and H3 really means just sharing L7 filter chain config and possibly transport socket. But I don't know how this can be easily achieved.